### ISSUE TITLE
Add type hinting

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/py_ecc#11]
> Largely copy/pasted from https://github.com/ethereum/py-evm/issues/1398

### Background

[Type hints](https://www.python.org/dev/peps/pep-0484/) allow us to perform static type checking, among other things. They raise the security bar by catching bugs at development time that, without type support, may turn into runtime bugs.

[This stackoverflow answer](https://stackoverflow.com/questions/32557920/what-are-type-hints-in-python-3-5/32558710#32558710) does a great job at describing their main benefits.

### What is wrong?

This library currently does not have any type hints.

This needs to be fixed by:

1. Adding all missing type hints.
2. Enforcing (stricter) type checking in CI

### How

There does exist tooling ([monkeytype](https://engineering.instagram.com/let-your-code-type-hint-itself-introducing-open-source-monkeytype-a855c7284881)) to the generation of type hints for existing code bases. From my personal experience `monkeytype` *can* be helpful but does still require manual fine tuning. Also, manually adding these type hints does serve as a great boost to the general understanding of the code base as it forces one to think about the code.

1. Run `mypy --follow-imports=silent --warn-unused-ignores --ignore-missing-imports --no-strict-optional --check-untyped-defs --disallow-incomplete-defs --disallow-untyped-defs --disallow-any-generics -p eth`

2. Eliminate every reported error by adding the right type hint

Because this library supports older versions of python, the type hints will not be able to use the modern python3.6 syntax.

### Definition of done

This issue is done when the following criteria are met:

1. `mypy` is run in CI

Add a new command to the `flake8` environment in the `tox.ini` file that runs:

```
mypy --follow-imports=silent --warn-unused-ignores --ignore-missing-imports --no-strict-optional --check-untyped-defs --disallow-incomplete-defs --disallow-untyped-defs --disallow-any-generics -p py_ecc`
```

2. Usage of `type: ignore` (silencing the type checker) is minimized and there's a reasonable explanation for its usage 

### Stretch goals

When this issue is done, stretch goals can be applied (and individually get funded) to tighten type support to qualify:

1. `mypy --strict --follow-imports=silent --ignore-missing-imports --no-strict-optional -p py_ecc`

**Comment by 6ug** [comment]
I like to work on this thing, I just saw talk from guido and I can do this thing.

**Comment by pipermerriam** [comment]
@6ug :+1: 

Please get a pull request open early (like as soon as you have any changes committed).  We'll mark it as a work in progress while you complete it.

**Comment by pipermerriam** [comment]
Also, you should just include the necessary changes for #12 in this as well as they don't really need to be done separately.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 360.0 DAI (360.0 USD @ $1.0/DAI)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/py_ecc/11/1531).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ethereum/py_ecc/11/1531).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $48,817.80 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by iamonuwa** [comment]
@6ug are you interested in this? Am ready to start working on it immediately

**Comment by iamonuwa** [comment]
I'll apply for this. But if you are interested, you can work on it.

**Comment by 6ug** [comment]
hey hi @iamonuwa, sorry I did not saw your comment. Yeah, I like to work on this, may be we can work more collaboratively. You can start solving it by end of `mypy` output and I will start solving by top, will that be okay?

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 360.0 DAI (360.0 USD @ $1.0/DAI)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/py_ecc/11/1532).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ethereum/py_ecc/11/1532).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $49,073.89 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 2 months, 4 weeks from now.
Please review their action plans below:


**1) [svenski123](https://gitcoin.co/profile/svenski123) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/ethereum/py_ecc/11/1531?mutate_worker_action=approve&worker=svenski123) | [reject worker](https://gitcoin.co/issue/ethereum/py_ecc/11/1531?mutate_worker_action=reject&worker=svenski123))_.**

Hi, happy to do this one but please confirm it is the py_ecc package in https://github.com/ethereum/py_ecc.git that needs the type hinting (which is what "View on Github" links to).

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/py_ecc/11/1531).

**Comment by svenski123** [comment]
Hi the type hinting issue that I applied to work was opened about 30 minutes ago and so I jumped it as I was late to the original py-evm.eth bounty. However it seems there is already a long github thread in place which is linked to multiple issues/bounties so I am not sure if this has been taken or not. Please kindly clarify which modules (if any) are to be worked on by whom.

**Comment by ceresstation** [comment]
Hey @iamonuwa @6ug did you both apply? I don't see you in the list of applicants.

**Comment by pipermerriam** [comment]
We generally do fixst-come-first-serve with the expectation that when there are multiple parties interested in doing a task that the work will be done *promptly* by the person claiming (loosely defined as PR open within the first 24 hours and actively working on the issue).

@6ug has first dibs, followed by @iamonuwa 

If the two of you (@6ug and @iamonuwa ) want to team up on this, make sure you two have a clear and public agreement on how bounty payout will work if you are both interested in the bounty).

**Comment by pipermerriam** [comment]
@svenski123 yes, py_ecc package, apologies if the gitcoin issue says something different, I was creating a bunch of these across multiple repositories.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 11 months, 4 weeks from now.
Please review their action plans below:


**1) [federicosan](https://gitcoin.co/profile/federicosan) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?mutate_worker_action=approve&worker=federicosan) | [reject worker](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?mutate_worker_action=reject&worker=federicosan))_.**

mypy --follow-imports=silent --warn-unused-ignores --ignore-missing-imports --                                                                                                                                  no-strict-optional --check-untyped-defs --disallow-incomplete-defs --disallow-un                                                                                                                                  typed-defs --disallow-any-generics -p py_ecc
py_ecc\secp256k1\secp256k1.py:9: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:26: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:34: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:46: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:51: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:63: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:87: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:92: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:105: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:109: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:113: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:117: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:128: error: Function is missing a type annotation
py_ecc\secp256k1\secp256k1.py:140: error: Function is missing a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:12: error: Name 'long' is not                                                                                                                                   defined
py_ecc\optimized_bn128\optimized_field_elements.py:14: error: Incompatible types                                                                                                                                   in assignment (expression has type "Tuple[Type[int]]", variable has type "Tuple                                                                                                                                  [Type[int], Any]")
py_ecc\optimized_bn128\optimized_field_elements.py:19: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:34: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:41: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:45: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:49: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:52: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:55: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:59: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:63: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:68: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:71: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:76: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:79: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:89: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:95: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:98: error: Function is missin                                                                                                                                  g a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:101: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:105: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:109: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:114: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:121: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:135: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:143: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:151: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:159: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:176: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:179: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:190: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:193: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:204: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:222: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:225: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:232: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:235: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:239: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:243: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:249: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\optimized_bn128\optimized_field_elements.py:259: error: Function is missi                                                                                                                                  ng a type annotation
py_ecc\bn128\bn128_field_elements.py:11: error: Name 'long' is not defined
py_ecc\bn128\bn128_field_elements.py:13: error: Incompatible types in assignment                                                                                                                                   (expression has type "Tuple[Type[int]]", variable has type "Tuple[Type[int], An                                                                                                                                  y]")
py_ecc\bn128\bn128_field_elements.py:27: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:42: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:49: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:53: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:57: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:60: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:63: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:67: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:71: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:76: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:79: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:84: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:87: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:97: error: Function is missing a type annot                                                                                                                                  ation
py_ecc\bn128\bn128_field_elements.py:103: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:106: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:109: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:113: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:117: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:122: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:129: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:141: error: Unsupported operand types for +                                                                                                                                   ("Tuple[Type[FQ]]" and "Tuple[Type[int], Any]")
py_ecc\bn128\bn128_field_elements.py:146: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:154: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:158: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:162: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:177: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:180: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:187: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:190: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:201: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:219: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:222: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:229: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:232: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:236: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:240: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:246: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\bn128\bn128_field_elements.py:255: error: Function is missing a type anno                                                                                                                                  tation
py_ecc\optimized_bn128\optimized_curve.py:46: error: Function is missing a type                                                                                                                                   annotation
py_ecc\optimized_bn128\optimized_curve.py:51: error: Function is missing a type                                                                                                                                   annotation
py_ecc\optimized_bn128\optimized_curve.py:63: error: Function is missing a type                                                                                                                                   annotation
py_ecc\optimized_bn128\optimized_curve.py:77: error: Function is missing a type                                                                                                                                   annotation
py_ecc\optimized_bn128\optimized_curve.py:105: error: Function is missing a type                                                                                                                                   annotation
py_ecc\optimized_bn128\optimized_curve.py:116: error: Function is missing a type                                                                                                                                   annotation
py_ecc\optimized_bn128\optimized_curve.py:122: error: Function is missing a type                                                                                                                                   annotation
py_ecc\optimized_bn128\optimized_curve.py:132: error: Function is missing a type                                                                                                                                   annotation
py_ecc\optimized_bn128\optimized_curve.py:139: error: Function is missing a type                                                                                                                                   annotation
py_ecc\bn128\bn128_curve.py:45: error: Function is missing a type annotation
py_ecc\bn128\bn128_curve.py:50: error: Function is missing a type annotation
py_ecc\bn128\bn128_curve.py:62: error: Function is missing a type annotation
py_ecc\bn128\bn128_curve.py:71: error: Function is missing a type annotation
py_ecc\bn128\bn128_curve.py:89: error: Function is missing a type annotation
py_ecc\bn128\bn128_curve.py:100: error: Function is missing a type annotation
py_ecc\bn128\bn128_curve.py:109: error: Function is missing a type annotation
py_ecc\bn128\bn128_curve.py:116: error: Function is missing a type annotation
py_ecc\optimized_bn128\optimized_pairing.py:36: error: Function is missing a typ                                                                                                                                  e annotation
py_ecc\optimized_bn128\optimized_pairing.py:45: error: Function is missing a typ                                                                                                                                  e annotation
py_ecc\optimized_bn128\optimized_pairing.py:69: error: Function is missing a typ                                                                                                                                  e annotation
py_ecc\optimized_bn128\optimized_pairing.py:98: error: Function is missing a typ                                                                                                                                  e annotation
py_ecc\optimized_bn128\optimized_pairing.py:138: error: Function is missing a ty                                                                                                                                  pe annotation
py_ecc\optimized_bn128\optimized_pairing.py:146: error: Function is missing a ty                                                                                                                                  pe annotation
py_ecc\bn128\bn128_pairing.py:27: error: Function is missing a type annotation
py_ecc\bn128\bn128_pairing.py:42: error: Function is missing a type annotation
py_ecc\bn128\bn128_pairing.py:71: error: Function is missing a type annotation
py_ecc\bn128\bn128_pairing.py:95: error: Function is missing a type annotation
py_ecc\bn128\bn128_pairing.py:101: error: Function is missing a type annotation

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/py_ecc/11/1532).

**Comment by federicosan** [comment]
Hi, I applied to work on this via gitcoin, please confirm is this is already taken.

**Comment by pipermerriam** [comment]
@federicosan work is already underway in https://github.com/ethereum/py_ecc/pull/18

**Comment by gitcoinbot** [comment]
@federicosan Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=100)

**Comment by gitcoinbot** [comment]
@federicosan Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@federicosan due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/ethereum/py_ecc/11/1532) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] warning (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@federicosan due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/ethereum/py_ecc/11/1532) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] warning (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/py_ecc/11/1532?snooze=100)

**Comment by Bhargavasomu** [comment]
@pipermerriam as I don't see progress in the issue (as per gitcoin), I would like to work on this. I can complete this in a day. Do I have the permission to go ahead with this?

**Comment by ryan-shea** [comment]
Hey @pipermerriam, what is the status on this? Would love to give @Bhargavasomu the opportunity to complete this for you. On Gitcoin, @federicosan is the only worker approved. How can I help here? 😄

**Comment by pipermerriam** [comment]
@ryan-shea just got back from some vacation post-devcon.  @Bhargavasomu has my blessing to work on this.

**Comment by Bhargavasomu** [comment]
@gitcoinbot could you please reopen the issue and later approve me?

**Comment by ceresstation** [comment]
You should be approved now @Bhargavasomu 👍

**Comment by ceresstation** [comment]
Hey @Bhargavasomu any updates on your end?

**Comment by Bhargavasomu** [comment]
@ceresstation I am in the middle of the examinations and have completed half of the work. Will work full time on this in around 5 days. Is that ok? (Already done half of the supportive work, though not in the state for a PR)

**Comment by Bhargavasomu** [comment]
@pipermerriam why are we using seperate classes for `FQ` when it is just like an `int`?

**Comment by pipermerriam** [comment]
@Bhargavasomu this was all originally written by Vitalik and I don't actually have a strong understanding of the internals.  I'd shoot for absolute minimal changes to the code as part of this as I'm not willing to sign off on anything that does much in terms of structural changes that aren't easy to visually verify that they don't change any functionality.

**Comment by Bhargavasomu** [comment]
@pipermerriam I think that there are some potential problems in this code, when it comes to type hinting because of the design.

https://github.com/ethereum/py_ecc/blob/25150ae59eb5b29185e3607d48c9b391392fe3ee/py_ecc/bn128/bn128_field_elements.py#L41-L47
If you take a look at the above snippet, the type hinting for the function argument n, as per the code should be `Union[int, "FQ"]`. But the type `int` doesn't have the attribute `n` and hence mypy is also throwing the same error.
```
error: Item "int" of "Union[int, FQ]" has no attribute "n"
```
We could do a `type: ignore`, but there are a lot of places with similar design and would involve a bunch of `type: ignore` which would make the code messy. How should I go about this?

**Comment by Bhargavasomu** [comment]
@pipermerriam there are a lot of `assert` statements between the code, not as part of any function but as part of the main code itself.
https://github.com/ethereum/py_ecc/blob/25150ae59eb5b29185e3607d48c9b391392fe3ee/py_ecc/optimized_bn128/optimized_pairing.py#L76-L94
`mypy` apparently has some issues with it I guess, because it is giving irrelevant errors. Apart from `mypy`, it would be better if the `assert` statements would be placed in `tests` directory right? Could you guide me more on this as in where to place all these `assert` statements?

**Comment by Bhargavasomu** [comment]
@pipermerriam this issue is still not over as the type hinting of `optimized_bn128`. I will be submitting the PR for that soon though.

**Comment by pipermerriam** [comment]
https://github.com/ethereum/py_ecc/blob/master/tox.ini#L27-L29

@Bhargavasomu can you get in one more PR which condenses those three CI runs into a single run on the base module so that we can confirm that the whole library is properly typed/linted.

**Comment by Bhargavasomu** [comment]
@pipermerriam the main problem with that is the sub-modules `bs12_381` and `optimized_bs12_381` are not type hinted as of now; and hence `mypy` check on the base module(`py_ecc`) would fail. I could type hint it, but it would lead to a lot of redundancy and doing a refactoring first would help in the type hinting. I have added all the details [here](https://github.com/ethereum/py_ecc/issues/24#issuecomment-442701351). I will first try to tackle the refactoring part as of #24 and then finish the type hinting for the rest of the sub-modules.

**Comment by Bhargavasomu** [comment]
@pipermerriam I would like to take up #31. Once I am done with it, will come back to this.

**Comment by Bhargavasomu** [comment]
And by the way, could we have a blocked label for this issue?

**Comment by catslovesmilk** [comment]
Hey folks, can I take this issue?

**Comment by pipermerriam** [comment]
@catslovesmilk 

Somu (@Bhargavasomu ) has been working towards this for a bit and was blocked by his work on #31 so for now it is still his.

**Comment by Bhargavasomu** [comment]
@pipermerriam I think this bounty be closed with #63.

**Comment by pipermerriam** [comment]
@gitcoinbot is it possible to re-activate this bounty to be paid towards @Bhargavasomu 

This issue ended up being blocked/delayed by some other deeper work that Somu did on the library and it's just now been completed.

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 360.00000  DAI (360.0 USD @ $1.0/DAI) has been granted to @Bhargavasomu for this issue from @ceresstation. ⚡️ 

Nice work @Bhargavasomu! Your tip has automatically been deposited in the ETH address we have on file.

 * $53654.53 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__This Bounty has been completed.__

Additional Tips for this Bounty:
* ceresstation tipped 360.0000 DAI worth 360.0 USD to Bhargavasomu.

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/py_ecc/11/1531)
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $53,294.53 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by lazaridiscom** [comment]
has still "work started" stgatus on gitcoin

**Comment by lazaridiscom** [comment]
https://gitcoin.co/issue/ethereum/py_ecc/11/1532
